syntax = "proto3";

package university;

import "google/protobuf/empty.proto";

enum LoanStatus {
  ONGOING = 0;
  RETURNED = 1;
  LATE = 2;
}

message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 pageCount = 6;
  int32 stock = 7;
}

message Student {
  string id = 1;
  string name = 2;
  string studentNumber = 3;
  string email = 4;
  bool isActive = 5;
}

message Loan {
  string id = 1;
  string studentId = 2;
  string bookId = 3;
  string loanDate = 4;
  string returnDate = 5;
  LoanStatus status = 6;
}

message BookIdRequest { string id = 1; }
message StudentIdRequest { string id = 1; }
message LoanIdRequest { string id = 1; }

message BookList { repeated Book books = 1; }
message StudentList { repeated Student students = 1; }
message LoanList { repeated Loan loans = 1; }

service BookService {
  rpc GetBook(BookIdRequest) returns (Book);
  rpc ListBooks(google.protobuf.Empty) returns (BookList);
  rpc CreateBook(Book) returns (Book);
  rpc UpdateBook(Book) returns (Book);
  rpc DeleteBook(BookIdRequest) returns (google.protobuf.Empty);
}

service StudentService {
  rpc GetStudent(StudentIdRequest) returns (Student);
  rpc ListStudents(google.protobuf.Empty) returns (StudentList);
  rpc CreateStudent(Student) returns (Student);
  rpc UpdateStudent(Student) returns (Student);
  rpc DeleteStudent(StudentIdRequest) returns (google.protobuf.Empty);
}

service LoanService {
  rpc GetLoan(LoanIdRequest) returns (Loan);
  rpc ListLoans(google.protobuf.Empty) returns (LoanList);
  rpc BorrowBook(Loan) returns (Loan);
  rpc ReturnBook(LoanIdRequest) returns (Loan);
}
